biscuit-auth/src/datalog/expression.rs

```diff
@@ -205,16 +229,31 @@ impl Binary {
             // boolean
             (Binary::And, Term::Bool(i), Term::Bool(j)) => Ok(Term::Bool(i & j)),
             (Binary::Or, Term::Bool(i), Term::Bool(j)) => Ok(Term::Bool(i | j)),
-            (Binary::Equal, Term::Bool(i), Term::Bool(j)) => Ok(Term::Bool(i == j)),
-            (Binary::NotEqual, Term::Bool(i), Term::Bool(j)) => Ok(Term::Bool(i != j)),
+            (Binary::Equal | Binary::HeterogeneousEqual, Term::Bool(i), Term::Bool(j)) => {
+                Ok(Term::Bool(i == j))
+            }
+            (Binary::NotEqual | Binary::HeterogeneousNotEqual, Term::Bool(i), Term::Bool(j)) => {
+                Ok(Term::Bool(i != j))
+            }
 
             // null
-            (Binary::Equal, Term::Null, Term::Null) => Ok(Term::Bool(true)),
-            (Binary::Equal, Term::Null, _) => Ok(Term::Bool(false)),
-            (Binary::Equal, _, Term::Null) => Ok(Term::Bool(false)),
-            (Binary::NotEqual, Term::Null, Term::Null) => Ok(Term::Bool(false)),
-            (Binary::NotEqual, Term::Null, _) => Ok(Term::Bool(true)),
-            (Binary::NotEqual, _, Term::Null) => Ok(Term::Bool(true)),
+            (Binary::Equal | Binary::HeterogeneousEqual, Term::Null, Term::Null) => {
+                Ok(Term::Bool(true))
+            }
+            (Binary::Equal | Binary::HeterogeneousEqual, Term::Null, _) => Ok(Term::Bool(false)),
```

Comment:
since we're separating between implementations for equal, maybe null should always return an error in a strict comparison with any other type, for consistency. @divarvel WDYT?

Reply:
Oh, I missed that, good catch.

Reply:
```suggestion
            (Binary::HeterogeneousEqual, Term::Null, _) => Ok(Term::Bool(false)),
```

(same for `NotEqual`)